server/tso/tso.go

```diff
@@ -215,20 +218,21 @@ func (t *timestampOracle) UpdateTimestamp(leadership *election.Leadership) error
 	}
 
 	var next time.Time
-	prevLogical := atomic.LoadInt64(&prev.logical)
+	prevLogical := t.tso.logical
 	// If the system time is greater, it will be synchronized with the system time.
 	if jetLag > updateTimestampGuard {
 		next = now
 	} else if prevLogical > maxLogical/2 {
 		// The reason choosing maxLogical/2 here is that it's big enough for common cases.
 		// Because there is enough timestamp can be allocated before next update.
 		log.Warn("the logical time may be not enough", zap.Int64("prev-logical", prevLogical))
-		next = prev.physical.Add(time.Millisecond)
+		next = t.tso.physical.Add(time.Millisecond)
 	} else {
 		// It will still use the previous physical time to alloc the timestamp.
 		tsoCounter.WithLabelValues("skip_save").Inc()
 		return nil
 	}
+	t.tsoMux.RUnlock()
```

Comment:
When not use *defer unlock*, it is a good practice to make unlock near lock as much as possible. A common mistake is forget to unlock the lock when there are early returns. (line 233 in this case)